com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs

```diff
@@ -13,23 +14,32 @@ namespace Unity.MLAgents.Inference
     /// </summary>
     internal class ContinuousActionOutputApplier : TensorApplier.IApplier
     {
-        public void Apply(TensorProxy tensorProxy, IEnumerable<int> actionIds, Dictionary<int, float[]> lastActions)
+        readonly ActionSpec m_ActionSpec;
+
+        public ContinuousActionOutputApplier(ActionSpec actionSpec)
+        {
+            m_ActionSpec = actionSpec;
+        }
+
+        public void Apply(TensorProxy tensorProxy, IEnumerable<int> actionIds, Dictionary<int, ActionBuffers> lastActions)
         {
             var actionSize = tensorProxy.shape[tensorProxy.shape.Length - 1];
             var agentIndex = 0;
             foreach (int agentId in actionIds)
             {
                 if (lastActions.ContainsKey(agentId))
                 {
-                    var actionValue = lastActions[agentId];
-                    if (actionValue == null)
+                    var actionBuffer = lastActions[agentId];
+                    if (actionBuffer.Equals(ActionBuffers.Empty))
                     {
-                        actionValue = new float[actionSize];
-                        lastActions[agentId] = actionValue;
+                        actionBuffer = new ActionBuffers(new float[m_ActionSpec.NumContinuousActions],
```

Comment:
What I don't like about this is ContinuousActionOutputApplier has to initialize new ActionBuffers based on discrete actions branchSizes.

Reply:
Maybe make another ActionBuffers constructor (or static method) that takes an ActionSpec?

Reply:
Yes I can do that.
But even with that I'm still not able to initialize continuous/discrete field separately and still need both sizes to initialize the new ActionBuffers here, due to ActionBuffers's read-only properties. I wonder if it's reasonable to change that.

Reply:
I don't think initializing them separately is a requirement; you have all the information you need in the ActionSpec. Encapsulating it in an ActionBuffers static method seems fine to me.

You could also move the logic to a ModelRunner method, and call it in ModelRunner.DecideBatch(), just before the call to m_TensorApplier.ApplyTensors.